lib/lambda/update/updatePackage.ts

```diff
@@ -0,0 +1,232 @@
+import { response } from "libs/handler-lib";
+import { APIGatewayEvent } from "aws-lambda";
+import { getPackage } from "libs/api/package";
+import { produceMessage } from "libs/api/kafka";
+import { ItemResult } from "shared-types/opensearch/main";
+import { getPackageType } from "./getPackageType";
+import { events } from "lib/packages/shared-types";
+import { z } from "zod";
+
+const sendDeleteMessage = async (packageId: string) => {
+  const topicName = process.env.topicName as string;
+  if (!topicName) {
+    throw new Error("Topic name is not defined");
+  }
+  await produceMessage(
+    topicName,
+    packageId,
+    JSON.stringify({
+      id: packageId,
+      deleted: true,
+      isAdminChange: true,
+      adminChangeType: "delete",
+    }),
+  );
+
+  return response({
+    statusCode: 200,
+    body: { message: `${packageId} has been deleted.` },
+  });
+};
+
+const sendUpdateValuesMessage = async ({
+  currentPackage,
+  updatedFields,
+  changeReason,
+}: {
+  currentPackage: ItemResult;
+  updatedFields: object;
+  changeReason?: string;
+}) => {
+  const topicName = process.env.topicName as string;
+  if (!topicName) {
+    throw new Error("Topic name is not defined");
+  }
+  const invalidFields = Object.keys(updatedFields).filter(
+    (field) => !(field in currentPackage._source),
+  );
+  if (invalidFields.length > 0) {
+    return response({
+      statusCode: 400,
+      body: { message: `Cannot update invalid field(s): ${invalidFields.join(", ")}` },
+    });
+  }
+
+  if ("id" in updatedFields) {
+    return response({
+      statusCode: 400,
+      body: { message: "ID is not a valid field to update" },
+    });
+  }
+
+  const fieldNames = Object.keys(updatedFields).join(", ");
+  const changeMadeText = `${fieldNames} ${
+    Object.keys(updatedFields).length > 1 ? "have" : "has"
+  } been updated`;
+
+  await produceMessage(
+    topicName,
+    currentPackage._id,
+    JSON.stringify({
+      id: currentPackage._id,
+      ...updatedFields,
+      isAdminChange: true,
+      adminChangeType: "update-values",
+      changeMade: changeMadeText,
+      changeReason,
+    }),
+  );
+
+  return response({
+    statusCode: 200,
+    body: { message: `${changeMadeText} in package ${currentPackage._id}.` },
+  });
+};
+
+const sendUpdateIdMessage = async ({
+  currentPackage,
+  updatedId,
+}: {
+  currentPackage: ItemResult;
+  updatedId: string;
+}) => {
+  const topicName = process.env.topicName as string;
+  if (!topicName) {
+    throw new Error("Topic name is not defined");
+  }
+  // ID and changeMade are excluded; the rest of the object has to be spread into the new package
```

Comment:
Seems like `origin` is also excluded?

Reply:
I'm actually using origin and had to set it to OneMAC a little below